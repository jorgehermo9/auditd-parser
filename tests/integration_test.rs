use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
