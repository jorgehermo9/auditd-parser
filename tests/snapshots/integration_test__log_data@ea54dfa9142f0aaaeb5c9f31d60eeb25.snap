---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/linux-audit-parser-rs/line-acct-lock.log:1"
expression: result
info: "type=ACCT_LOCK msg=audit(1725000411.409:1065): pid=2352 uid=0 auid=4294967295 ses=4294967295 msg='op=locked-password id=1000 exe=\"/usr/bin/passwd\" hostname=? addr=? terminal=? res=success'\u001dUID=\"root\" AUID=\"unset\" ID=\"ec2-user\""
---
{
  "record_type": "ACCT_LOCK",
  "timestamp": 1725000411409,
  "id": 1065,
  "fields": {
    "auid": null,
    "msg": {
      "addr": null,
      "exe": "/usr/bin/passwd",
      "hostname": null,
      "id": 1000,
      "op": "locked-password",
      "res": "success",
      "terminal": null
    },
    "pid": "2352",
    "ses": "4294967295",
    "uid": "root"
  },
  "enrichment": {
    "AUID": "unset",
    "ID": "ec2-user",
    "UID": "root"
  }
}
